### PR TITLE
fwupd-efi: init at 1.1, fwupd: 1.5.12 -> 1.7.1

### PR DESCRIPTION
**Pull request by amaxine**
###### Motivation for this change
Changelogs:
- https://github.com/fwupd/fwupd/releases/tag/1.6.0
- https://github.com/fwupd/fwupd/releases/tag/1.6.1
- https://github.com/fwupd/fwupd/releases/tag/1.6.2
- https://github.com/fwupd/fwupd/releases/tag/1.6.3
- https://github.com/fwupd/fwupd/releases/tag/1.6.4
- https://github.com/fwupd/fwupd/releases/tag/1.7.0
- https://github.com/fwupd/fwupd/releases/tag/1.7.1

Announcement: https://blogs.gnome.org/hughsie/2021/04/26/fwupd-1-6-0/

There is some overlap with 1.5.x for those changes as 1.6.0 came out in April already.

Opening as a draft because it fails on the install step (seemingly in `meson_post_install.sh`), but I did not have the time to properly look into the "why" yet.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch [hunk 1]
```diff
@@ -4,5 +4,5 @@ index c7a430c0..e69de29b 100644
 +++ b/data/builder/meson.build
 @@ -1,3 +0,0 @@
 -install_data('README.md',
--  install_dir : join_paths(localstatedir, 'lib', 'fwupd', 'builder')
+-  install_dir : join_paths(datadir, 'doc', 'fwupd', 'builder')
 -)
```
Comment (jtojnar) [inline pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch:8]: Is this patch necessary? Writing to datadir should not cause issues.

File: pkgs/os-specific/linux/firmware/fwupd/default.nix [hunk 2]
```diff
@@ -165,12 +166,13 @@ let
       pango
       tpm2-tss
       efivar
```
Comment (jtojnar) [inline pkgs/os-specific/linux/firmware/fwupd/default.nix:168]: `cairo`, `freetype` and `fontconfig` might not be needed any more.

[`1.5.12...1.6.0`](https://github.com/fwupd/fwupd/compare/1.5.12...1.6.0#diff-30d8f6be6320feeacf686be94f48c70869b52630e01ea625f0f15adc0d57c3e4L349-L355)

File: pkgs/os-specific/linux/firmware/fwupd-efi/default.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
+{ lib, stdenv
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/firmware/fwupd-efi/default.nix:1]:
```suggestion
{ lib
, stdenv
```

File: pkgs/os-specific/linux/firmware/fwupd/default.nix [hunk 4]
```diff
@@ -91,16 +92,16 @@ let
...
 
     # libfwupd goes to lib
     # daemon, plug-ins and libfwupdplugin go to out
     # CLI programs go to out
     outputs = [ "out" "lib" "dev" "devdoc" "man" "installedTests" ];
 
     src = fetchurl {
-      url = "https://people.freedesktop.org/~hughsient/releases/fwupd-${version}.tar.xz";
-      sha256 = "sha256-BluwLlm6s/2H/USARQpAvDR0+X8WP/q0h8VvxA6Qftc=";
+      url = "https://people.freedesktop.org/~hughsient/releases/${pname}-${version}.tar.xz";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/firmware/fwupd/default.nix:103]: We shouldn't change URLs to contain more variables then necessary. It makes it easier to compare them to upstream ones or manually visit them.

File: pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch (hunk 1 above)
Comment (maxeaubrey) [reply pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch:8]: Ack, wasn't thinking when I opened this. I agree it doesn't seem necessary, will fix it up.

File: pkgs/os-specific/linux/firmware/fwupd/default.nix (hunk 2 above)
Comment (maxeaubrey) [reply pkgs/os-specific/linux/firmware/fwupd/default.nix:168]: Thanks for spotting! Correct on all 5, of course.

**Comment by maxeaubrey** [comment]
Thanks for the reviews. I'll hopefully have time this week to actually look at why the install step fails.

**Comment by jtojnar** [comment]
Using

```patch
--- a/pkgs/os-specific/linux/firmware/fwupd/default.nix
+++ b/pkgs/os-specific/linux/firmware/fwupd/default.nix
@@ -221,6 +221,10 @@ let
       addToSearchPath XDG_DATA_DIRS "${shared-mime-info}/share"
     '';
 
+    installPhase = ''
+      meson install --dry-run
+    '';
+
     postInstall =
       let
         testFw = fetchFromGitHub {
```

it says

```
Installing /build/fwupd-1.6.3/build/plugins/uefi-capsule/35_fwupd to /etc/grub.d
```

**Comment by maxeaubrey** [comment]
@jtojnar Thank you for the pointer! Sorry for taking so long tho, I did not have time at all after opening this PR :sweat_smile:

**Comment by jtojnar** [comment]
Looks like there is bug in fwupd, it hardcodes build directory in installed tests:

```
FuPluginWacomUsb:ERROR:../plugins/wacom-usb/fu-self-test.c:68:fu_wac_firmware_xml_func: assertion failed (error == NULL): Failed to open file ?/build/fwupd-1.6.3/src/fuzzing/wacom.builder.xml?: No such file or directory (g-file-error-quark, 4)
Bail out! FuPluginWacomUsb:ERROR:../plugins/wacom-usb/fu-self-test.c:68:fu_wac_firmware_xml_func: assertion failed (error == NULL): Failed to open file ?/build/fwupd-1.6.3/src/fuzzing/wacom.builder.xml?: No such file or directory (g-file-error-quark, 4)
```

**Comment by maxeaubrey** [comment]
Ok, so hopefully I won't forget to run tests every single time in the future as I have so far.

Looking more closely, I /think/ we need to fix things around `FWUPD_FUZZINGSRCDIR` rather than `TESTDATADIR_DST`, as that's the (hardcoded) path `wacom-usb-self-test` and `synaptics-prometheus-self-test` fail on. There's also `redfish-self-test` that fails in the default config, as we do not build redfish support normally, but that's not related to those failures.

I'm not sure how to fix that up. Should I be installing the fuzzing files as part of the `installedTests` output and patching https://github.com/fwupd/fwupd/blob/1210aa4ae7a8bddfe4d0071736c1285f17136faa/meson.build#L511 ?

However, even after disabling those tests, I see a different failure that I don't really know where to start on:
```
Failed to create file “/etc/fwupd/remotes.d/fwupd-tests.conf.VSML90”: Read-only file system
-- Journal begins at Wed 2021-09-08 14:07:28 UTC, ends at Wed 2021-09-08 14:07:52 UTC. --
Sep 08 14:07:49 machine systemd[1]: Starting Firmware update daemon...
Sep 08 14:07:51 machine fwupd[934]: 14:07:51:0162 FuEngine             invalid has incorrect built version invalid
Sep 08 14:07:52 machine systemd[1]: Started Firmware update daemon.
FAIL: fwupd/fwupdmgr.test (Child process exited with code 1)
SUMMARY: total=2; passed=1; skipped=0; failed=1; user=0.2s; system=0.5s; maxrss=15772
FAIL: fwupd/fwupdmgr.test (Child process exited with code 1)

Traceback (most recent call last):
  File "/nix/store/9mbn9r1a3pcjb03x5rmnq48hka91p1y4-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 1096, in <module>
    run_tests(interactive)
  File "/nix/store/9mbn9r1a3pcjb03x5rmnq48hka91p1y4-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 933, in run_tests
    test_script()
  File "/nix/store/9mbn9r1a3pcjb03x5rmnq48hka91p1y4-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 1092, in test_script
    exec(testscript, symbols, None)
  File "<string>", line 1, in <module>
  File "/nix/store/9mbn9r1a3pcjb03x5rmnq48hka91p1y4-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 482, in succeed
    raise Exception(
Exception: command `gnome-desktop-testing-runner  -d '/nix/store/0bp0mrgwa103z4hykz0q4kn70xr8k2mb-fwupd-1.6.3-installedTests/share'` failed (exit code 2)
cleaning up
killing machine (pid 9)
(0.00 seconds)
builder for '/nix/store/slkgjvdv5l6amsrgxms2gfxnxxl6c0d7-vm-test-run-fwupd-1.6.3.drv' failed with exit code 1
```

Do you have any suggestions?

**Comment by jtojnar** [comment]
> Looking more closely, I /think/ we need to fix things around `FWUPD_FUZZINGSRCDIR` rather than `TESTDATADIR_DST`, as that's the (hardcoded) path `wacom-usb-self-test` and `synaptics-prometheus-self-test` fail on. There's also `redfish-self-test` that fails in the default config, as we do not build redfish support normally, but that's not related to those failures.

The issue is that we want the tests to work both at build time and when installed. (Well, we do not run tests at build time but upstream does so we need that in order for the patch to be upstreamable.)

The proper solution likely involves [`g_test_build_filename` function](https://docs.gtk.org/glib/func.test_build_filename.html) and yes, the files will need to be installed. (We should probably add CI workflow that runs installed tests upstream, since this is not the first time the installed tests were broken.)


> However, even after disabling those tests, I see a different failure that I don't really know where to start on:
Only the lines above `FAIL: fwupd/fwupdmgr.test (Child process exited with code 1)` are relevant. 

Passing `testRunnerFlags = "fwupd/fwupdmgr.test";` to `makeInstalledTest` in `nixos/tests/installed-tests/fwupd.nix` might reduce the noise.

**Comment by jtojnar** [comment]
Patched `fu_remote_list_set_key_value` and tests pass now.

**Review by jtojnar — APPROVED** [review]
Should be ready now.

**Pull request merged**